x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java

```diff
@@ -1711,24 +1706,27 @@ private void maybeStartTokenRemover() {
      * Gets the token from the <code>Authorization</code> header if the header begins with
      * <code>Bearer </code>
      */
-    private String getFromHeader(ThreadContext threadContext) {
+    public SecureString extractBearerTokenFromHeader(ThreadContext threadContext) {
         String header = threadContext.getHeader("Authorization");
         if (Strings.hasText(header) && header.regionMatches(true, 0, "Bearer ", 0, "Bearer ".length())
             && header.length() > "Bearer ".length()) {
-            return header.substring("Bearer ".length());
+            char[] chars = new char[header.length() - "Bearer ".length()];
+            header.getChars("Bearer ".length(), header.length(), chars, 0);
+            return new SecureString(chars);
         }
         return null;
     }
 
     String prependVersionAndEncodeAccessToken(Version version, String accessToken) throws IOException, GeneralSecurityException {
         if (version.onOrAfter(VERSION_ACCESS_TOKENS_AS_UUIDS)) {
-            try (ByteArrayOutputStream os = new ByteArrayOutputStream(MINIMUM_BASE64_BYTES);
-                 OutputStream base64 = Base64.getEncoder().wrap(os);
-                 StreamOutput out = new OutputStreamStreamOutput(base64)) {
+            try (BytesStreamOutput out = new BytesStreamOutput(MINIMUM_BASE64_BYTES)) {
                 out.setVersion(version);
                 Version.writeVersion(version, out);
+                if (version.onOrAfter(VERSION_TOKEN_TYPE)) {
+                    SecurityTokenType.ACCESS_TOKEN.write(out);
+                }
                 out.writeString(accessToken);
-                return new String(os.toByteArray(), StandardCharsets.UTF_8);
+                return Base64.getEncoder().encodeToString(out.bytes().toBytesRef().bytes);
```

Comment:
This is the issue that I burned a lot time on. The original way of serialisation does not work when you add an additional enum (vint) in between the existing version (vint) and string fields.

Reply:
Thanks to @tvernum We found that the issue can be fixed by close the base64 encoder stream. This is because the encoder otherwise does not know whether there will be more bytes coming and hence can hold bytes before they are multiple of 3. The original code worked because the byte length is 36 which happens to be a multiple of 3.